// element-plus/es/utils/props.js
/**
 * @description Build prop. It can better optimize prop types
 * @description 生成 prop，能更好地优化类型
 * @example
  // limited options
  // the type will be PropType<'light' | 'dark'>
  buildProp({
    type: String,
    values: ['light', 'dark'],
  } as const)
  * @example
  // limited options and other types
  // the type will be PropType<'small' | 'medium' | number>
  buildProp({
    type: [String, Number],
    values: ['small', 'medium'],
    validator: (val: unknown): val is number => typeof val === 'number',
  } as const)
  @link see more: https://github.com/element-plus/element-plus/pull/3341
 */
export function buildProp({ values, required, default: defaultValue, type, validator, } = {}) {
    return {
        type: type,
        required: !!required,
        default: defaultValue,
        validator: values || validator
            ? (val) => {
                let valid = false;
                if (values)
                    valid || (valid = [...values, defaultValue].includes(val));
                if (validator)
                    valid || (valid = validator(val));
                return valid;
            }
            : undefined,
    };
}
export const keyOf = (arr) => Object.keys(arr);
export const mutable = (val) => val;
export const componentSize = ['large', 'medium', 'small', 'mini'];
